src/core/smp.cc

```diff
@@ -218,6 +245,18 @@ internal::memory_prefaulter::work(std::vector<memory::internal::memory_range>& r
         // atomic_ref would be better, but alas C++20 only
         auto p1 = reinterpret_cast<volatile std::atomic<char>*>(p);
         p1->fetch_or(0, std::memory_order_relaxed);
+#endif
+#ifdef MADV_COLLAPSE
+    if (huge_page_size_opt) {
+        auto addr = reinterpret_cast<uintptr_t>(p);
+        auto aligned = addr & ~(page_size - 1);
+        auto offset_in_hugepage = aligned & (*huge_page_size_opt - 1);
+        auto first_page_offset = forwards ? 0 : *huge_page_size_opt - page_size;
+        auto hugepage_addr = aligned - offset_in_hugepage;
```

Comment:
Why not just `hugepage_addr = addr & ~(huge_page_size_opt - 1)` ?

Reply:
@avikivity , 

> Why not just hugepage_addr = addr & ~(huge_page_size_opt - 1) ?

Reply:
That just happened to be the way I wrote it. I'll reconsider, but more important is to use MAP_POPULATE_WRITE instead.